firedrake/assemble.py

```diff
@@ -277,12 +381,79 @@ def get_matrix(expr, mat_type, sub_mat_type, *, bcs=None,
         raise ValueError("Monolithic matrix assembly not supported for systems "
                          "with R-space blocks")
 
-    tensor = matrix.Matrix(expr, bcs, mat_type, sparsity, ScalarType,
-                           options_prefix=options_prefix)
-    return tensor, (), lambda: tensor
+    return matrix.Matrix(expr, bcs, opts.mat_type, sparsity, ScalarType,
+                         options_prefix=opts.options_prefix)
+
+
+def _assemble_expr(expr, tensor, bcs, opts, assembly_rank):
+    """Assemble an expression.
 
+    :arg expr: The expression to be assembled.
+    :arg tensor: The tensor to write to.
+    :arg bcs: Iterable of boundary conditions.
+    :arg opts: :class:`_AssemblyOpts` containing the assembly options.
+    :arg assembly_rank: The appropriate :class:`_AssemblyRank`.
+    """
+    # We cache the parloops on the form but keep track of the tensor. If the
+    # tensor is changed then we need to regenerate the parloop (until parloops
+    # stop depending on data).
+    # When we generate a new parloop the old one is removed from the cache to
+    # prevent leaking memory.
+    parloops = None
+    if isinstance(expr, ufl.form.Form):
+        cache_key = "parloops"
+        if cache_key in expr._cache:
+            cached_tensor, cached_parloops = expr._cache[cache_key]
+            if cached_tensor is tensor:
```

Comment:
Maybe we should check `tensor.bcs == bcs` here just in case.

Reply:
I don't think that would work because `tensor` doesn't have a `bcs` property. There is definitely a debate to be had here though about how safe this caching is. At the moment the cache will only miss if either the form or tensor are changed, not if `bcs`, `mat_type` etc are modified. I've added a warning about this to the documentation and it is not a common use case (in my opinion) for users to be modifying these options when assembling the same form.
Alongside this refactoring I'm also looking into making the parloop agnostic to as many of the `kwargs` as possible so hopefully this should become less of an issue.

Reply:
>  I've added a warning about this to the documentation and it is not a common use case (in my opinion) for users to be modifying these options when assembling the same form.

This is, unfortunately, not a reasonable position.

For example, there are (or would be after these changes) situations in the Firedrake source tree right now where the same form is (legitimately) assembled with different `mat_type` arguments. e.g. `AssembledPC` takes an `assemble(a, mat_type="matfree")` and assembles it with `assemble(a, mat_type="aij")`.

Caching is only ever allowed to be a performance optimisation, it isn't allowed to change behaviour.

Reply:
Ok. Thanks for the advice. I'll do some more thinking then about how to make the caching safe.

Reply:
This should now be resolved. I've checked and the performance hit from this alteration is negligible.

Reply:
I have now run all of the Thetis tests with this branch and they passed. I think the caching is sufficiently robust now.